File: 13 sorting/mergeSort.cs
  public static class MergeSorter
  {
    public static int[] MergeSort(int[] nums)
    {
      // if length > 2 then we need to divide
      if (nums.Length > 2)
      {
        double count = nums.Length / 2.0;

        // divide, leave remaining for the right side
        int[] leftSide = nums.Take((int)Math.Floor(count)).ToArray();
        int[] rightSide = nums.Skip(leftSide.Length).ToArray();

        // apply sorting to the divided
        leftSide = MergeSort(leftSide);
        rightSide = MergeSort(rightSide);

        // sort values from two sides into the main array
        int leftIndex = 0, rightIndex = 0;
        for (int i = 0; i < nums.Length; i++)
        {
          int? leftValue = leftIndex < leftSide.Length ? leftSide[leftIndex] : new int?();
          int? rightValue = rightIndex < rightSide.Length ? rightSide[rightIndex] : new int?();

          // if we finished the right side, or the left value is <= right
          if (rightValue == null || leftValue <= rightValue)
          {
            nums[i] = leftValue.Value;
            leftIndex++;
          }
          // if we finished the left side or the right value is < left
          else if (leftValue == null || rightValue < leftValue)
          {
            nums[i] = rightValue.Value;
            rightIndex++;
          }
        }
      }
      else if (nums.Length == 2)// length == 2, no need to divide
      {
        // sort
        if (nums[0] > nums[1])
        {
          int tmpValue = nums[0];
          nums[0] = nums[1];
          nums[1] = tmpValue;
        }
      }

      return nums;
    }
  }

File: 13 sorting/mergeSort.js
function mergeSort(nums) {
  // if length > 2 then we need to divide
  if (nums.length > 2) {
    let count = nums.length / 2.0;

    // divide, leave remaining for the right side
    let leftSide = nums.slice(0, Math.floor(count));
    let rightSide = nums.slice(leftSide.length);

    // apply sorting to the divided
    leftSide = mergeSort(leftSide);
    rightSide = mergeSort(rightSide);

    // sort values from two sides into the main array
    let leftIndex = 0, rightIndex = 0;
    for (let i = 0; i < nums.length; i++) {
      let leftValue = leftSide[leftIndex];
      let rightValue = rightSide[rightIndex];

      // if we finished the right side, or the left value is <= right
      if (rightValue == undefined || leftValue <= rightValue) {
        nums[i] = leftValue;
        leftIndex++;
      }
      // if we finished the left side or the right value is < left
      else if (leftValue == undefined || rightValue < leftValue) {
        nums[i] = rightValue;
        rightIndex++;
      }
    }
  }
  else if (nums.length == 2) // length == 2, no need to divide
  {
    // sort
    if (nums[0] > nums[1]) {
      let tmpValue = nums[0];
      nums[0] = nums[1];
      nums[1] = tmpValue;
    }
  }

  return nums;
}

let arr = [6, 5, 3, 1, 8, 7, 2, 4, 0];
console.log(mergeSort(arr));
